core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,424 @@
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut CtxStorage::access(ctx, &self.inner.pool)
+            .await
+            .wrap("access()")?;
+        let justification = storage
+            .certificate(ctx, number)
+            .await
+            .wrap("certificate()")?
+            .context("not found")?;
+        let payload = storage
+            .payload(ctx, number, self.inner.operator_address)
+            .await
+            .wrap("payload()")?
+            .context("miniblock disappeared from storage")?;
+        Ok(validator::FinalBlock {
+            payload: payload.encode(),
+            justification,
+        })
+    }
+
+    /// If actions queue is set (and the block has not been stored yet),
+    /// the block will be translated into a sequence of actions.
+    /// The received actions should be fed
+    /// to `ExternalIO`, so that `StateKeeper` will store the corresponding miniblock in the db.
+    ///
+    /// `store_next_block()` call will wait synchronously for the miniblock.
+    /// Once miniblock is observed in storage, `store_next_block()` will store a cert for this
+    /// miniblock.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()> {
+        // This mutex prevents concurrent store_next_block calls.
+        let mut cursor = ctx.wait(self.cursor.lock()).await?;
+        if let Some(cursor) = &mut *cursor {
+            let number = MiniblockNumber(
+                u32::try_from(block.header().number.0)
+                    .context("Integer overflow converting block number")?,
+            );
+            let payload =
+                Payload::decode(&block.payload).context("Failed deserializing block payload")?;
+            let want = cursor.inner.next_miniblock;
+            if want < number {
+                return Err(anyhow::anyhow!("expected {want:?}, got {number:?}").into());
+            }
+            if want == number {
+                let block = FetchedBlock {
+                    number,
+                    l1_batch_number: payload.l1_batch_number,
+                    last_in_batch: payload.last_in_batch,
+                    protocol_version: payload.protocol_version,
+                    timestamp: payload.timestamp,
+                    reference_hash: Some(payload.hash),
+                    l1_gas_price: payload.l1_gas_price,
+                    l2_fair_gas_price: payload.l2_fair_gas_price,
+                    virtual_blocks: payload.virtual_blocks,
+                    operator_address: payload.operator_address,
+                    transactions: payload.transactions,
+                };
+                cursor
+                    .actions
+                    .push_actions(cursor.inner.advance(block))
+                    .await;
```

Comment:
Nit: This can be extracted to a `Cursor` method taking `Payload` as an input to improve code structuring.

Reply:
done